pkg/operator/controller/enablednsnameresolver/controller.go

```diff
@@ -0,0 +1,130 @@
+package enablednsnameresolver
+
+import (
+	"context"
+	"sync"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	operatorcontroller "github.com/openshift/cluster-dns-operator/pkg/operator/controller"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "enable_dnsnameresolver_controller"
+)
+
+// New creates and returns a controller that starts the dependent caches
+// and controllers when the appropriate featuregate is enabled.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		cache:  mgr.GetCache(),
+		client: mgr.GetClient(),
+		config: config,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	clusterNamePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		expectedName := operatorcontroller.FeatureGateClusterConfigName()
+		actualName := types.NamespacedName{
+			Namespace: o.GetNamespace(),
+			Name:      o.GetName(),
+		}
+		return expectedName == actualName
+	})
+	if err := c.Watch(source.Kind(mgr.GetCache(), &configv1.FeatureGate{}), &handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DNSNameResolverEnabled indicates that the "DNSNameResolver" featuregate is enabled.
+	DNSNameResolverEnabled bool
+	// DependentCaches is a list of caches that are used by Controllers watching DNSNameResolver
+	// resources. The enable_dnsnameresolver controller starts these caches when the
+	// "DNSNameResolver" featuregate is enabled.
+	DependentCaches []cache.Cache
+	// DependentControllers is a list of controllers that watch DNSNameResolver
+	// resources. The enable_dnsnameresolver controller starts these controllers when the
+	// "DNSNameResolver" featuregate is enabled and the DependentCaches are started.
+	DependentControllers []controller.Controller
+}
+
+// reconciler handles the actual featuregate reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+
+	cache            cache.Cache
+	client           client.Client
+	startCaches      sync.Once
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a FeatureGate and starts the dependent
+// caches and controllers.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
```

Comment:
It's not clear to me why we need a Reconcile function for an operation we expect to only happen once.  It's misleading and could cause mistakes in the future.  Reconcile is for repeated actions we perform to reconcile updates or deletes as well as additions.  What does this reconcile once the feature gate is removed?  What happens if this is called more than once, and under what circumstance would it be called more than once?

Reply:
Previously the controller was ensuring the DNSNameResolver CRD in the reconcile loop. However, it was pointed out here (https://github.com/openshift/cluster-dns-operator/pull/394#discussion_r1497706122) that the CRDs can be placed in the manifests directory and the correct CRD depending on the feature set will picked up by CVO.

After removing the CRD part from the controller, it is now only starting the dnsnameresolver controller and the caches. It was also raised by @alebedev87 [previously](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029) that we remove the enablednsnameresolver controller completely and make dnsnameresolver controller managed. However, @Miciah suggested not to go for that change as we are close to the FF date. It is part of the future plan now to remove the enablednsnamresolver controller.

Reply:
Ok, if this will be removed in the future, just add a TODO comment explaining why/when, etc.

Reply:
Added the TODO in the operator.go before the dnsnameresolver controller is set up (https://github.com/openshift/cluster-dns-operator/pull/394/files#diff-99c0290799daf9abc6240df64063e20bfaf67b371577b67ac7eec6f4725622ffR148-R149).